File: app/src/main/java/com/brickgit/imagesearch/adapter/VideoAdapter.java
package com.brickgit.imagesearch.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.brickgit.imagesearch.R;
import com.brickgit.imagesearch.model.VideoInfoResponse;

import java.util.ArrayList;
import java.util.List;

public class VideoAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

	public interface OnVideoCellClickListener {
		void onVideoCellClicked(VideoInfoResponse videoInfo);
	}

	private List<VideoInfoResponse> listVideoInfo = new ArrayList<>();

	private VideoAdapter.OnVideoCellClickListener onVideoCellClickListener;

	private final View.OnClickListener onClickListener = new View.OnClickListener() {
		@Override
		public void onClick(View view) {
			if (onVideoCellClickListener != null) {
				VideoInfoResponse videoInfo = (VideoInfoResponse) view.getTag();
				if (videoInfo != null) {
					onVideoCellClickListener.onVideoCellClicked(videoInfo);
				}
			}
		}
	};

	public void setOnVideoCellClickListener(VideoAdapter.OnVideoCellClickListener onVideoCellClickListener) {
		this.onVideoCellClickListener = onVideoCellClickListener;
	}

	public void update(List<VideoInfoResponse> newVideos) {
		int size = listVideoInfo.size();
		listVideoInfo.clear();
		listVideoInfo.addAll(newVideos);
		notifyItemRangeInserted(size, newVideos.size() - size);
	}

	public void clear() {
		listVideoInfo.clear();
		notifyDataSetChanged();
	}

	@Override
	public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View view = LayoutInflater
				.from(parent.getContext())
				.inflate(R.layout.video_view_holder, null);
		view.setOnClickListener(onClickListener);
		return new VideoViewHolder(view);
	}

	@Override
	public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
		VideoInfoResponse videoInfo = listVideoInfo.get(position);
		VideoViewHolder viewHolder = (VideoViewHolder) holder;
		viewHolder.bind(videoInfo);
	}

	@Override
	public int getItemCount() {
		return listVideoInfo.size();
	}
}

File: app/src/main/java/com/guanhaolin/pearch/ui/video/VideoAdapter.kt
package com.guanhaolin.pearch.ui.video

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.guanhaolin.pearch.api.model.VideoResponse
import com.guanhaolin.pearch.databinding.VideoViewHolderBinding

class VideoAdapter : ListAdapter<VideoResponse, VideoViewHolder>(DIFF_CALLBACK) {
    interface OnVideoCellClickListener {
        fun onVideoCellClicked(video: VideoResponse)
    }

    private var onVideoCellClickListener: OnVideoCellClickListener? = null

    private val onClickListener =
        View.OnClickListener { view ->
            val video = view.tag as VideoResponse
            onVideoCellClickListener?.onVideoCellClicked(video)
        }

    fun setOnVideoCellClickListener(
        onVideoCellClickListener: OnVideoCellClickListener?
    ) {
        this.onVideoCellClickListener = onVideoCellClickListener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VideoViewHolder {
        val viewBinding =
            VideoViewHolderBinding.inflate(LayoutInflater.from(parent.context))
        viewBinding.root.setOnClickListener(onClickListener)
        return VideoViewHolder(viewBinding)
    }

    override fun onBindViewHolder(holder: VideoViewHolder, position: Int) {
        holder.bind(getItem(position))
    }
}

private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<VideoResponse>() {
    override fun areItemsTheSame(oldItem: VideoResponse, newItem: VideoResponse) =
        oldItem.id == newItem.id

    override fun areContentsTheSame(oldItem: VideoResponse, newItem: VideoResponse) =
        oldItem.views == newItem.views
                && oldItem.downloads == newItem.downloads
                && oldItem.likes == newItem.likes
                && oldItem.comments == newItem.comments
}
